Add vrf key to session keys type

4meta5 opened the pull request:
### What does it do?

Adds a new field to the runtime session keys type which corresponds to the VRF key type (assigned to `pallet_author_mapping::Keys` for now).  Also reassigns `Keys` associated type in AuthorMapping from `NimbusId` -> `VrfId`. This does not require any storage migration because both types are serialized the same way.

## Breaking Changes

⚠️ Runtime SessionKeys type changes from

```rust
pub struct SessionKeys {
	pub nimbus: AuthorInherent,
}
```

to 

```rust
pub struct SessionKeys {
	pub nimbus: AuthorInherent,
	pub vrf: session_keys_primitives::VrfSessionKey,
}
```

⚠️ `Keys` associated type for AuthorMapping changes from `NimbusId -> VrfId`. Both types are encoded the same way so no migration is required.

## Impact

This is a follow up of #1407 -- It ensures collators can call RPC method `author_rotateKeys` once instead of twice to generate the new keys required to rotate session keys.

librelois commented on primitives/session-keys/src/vrf.rs:
```diff
@@ -0,0 +1,67 @@
+	type Public = VrfId;
+}
+
+impl From<NimbusId> for VrfId {
+	fn from(nimbus_id: NimbusId) -> VrfId {
+		let nimbus_as_sr25519: sp_application_crypto::sr25519::Public = nimbus_id.into();
+		let sr25519_as_bytes: [u8; 32] = nimbus_as_sr25519.into();
+		sp_core::sr25519::Public::unchecked_from(sr25519_as_bytes).into()
+	}
+}
```
This code seems to me to be short and explicit enough, I don't see any point in shortening it, it would make it less readable.

librelois commented:
> This implies a storage migration is required for `MappingWithDeposit` map which has value type `RegistrationInfo` (which had the `keys` field change from `NimbusId` to `VrfId`).

A migration is necessary only if the serialized/encoded data must change.
And it seems to me that the NimbusId and VrfId types are encoded in the same way (in other words, they have the same serialized representation).
So, there is no need for migration.

4meta5 commented:
TODO: revert migration

librelois commented:
LGTM, it's a good work :+1:

crystalin commented:
> > This implies a storage migration is required for `MappingWithDeposit` map which has value type `RegistrationInfo` (which had the `keys` field change from `NimbusId` to `VrfId`).
> 
> A migration is necessary only if the serialized/encoded data must change. And it seems to me that the NimbusId and VrfId types are encoded in the same way (in other words, they have the same serialized representation). So, there is no need for migration.

If this PR is adding a new key type, (as the description is showing), how is that new key type going to be populated without a migration ?

crystalin commented:
@4meta5 also in the breaking change shouldn't we include the fact that the rotate_key is now returning the output of 2 keys instead of 1 ?

crystalin commented:
@4meta5 I don't understand why we changed the AuthorMapping key type from NimbusId to VrfId ? Isn't nimbus meant for authoring and Vrf meant for randomness support ?

4meta5 commented:
> If this PR is adding a new key type, (as the description is showing), how is that new key type going to be populated without a migration ?

It is already populated by the previous migration. The keys type in the AuthorMapping pallet was set equal to the NimbusId.

> also in the breaking change shouldn't we include the fact that the rotate_key is now returning the output of 2 keys instead of 1 ?

Yes, I added this.

> I don't understand why we changed the AuthorMapping key type from NimbusId to VrfId ? Isn't nimbus meant for authoring and Vrf meant for randomness support ?

The AuthorMapping key associated type is the key type for the per block VRF randomness support. We had previously set it equal to the actual NimbusId because we knew we are using the same key type so it will be serialized the same way when we make the switch here (so no migration will be necessary because they are serialized the same way).

crystalin commented:
Thank you @4meta5
